### ISSUE TITLE
Display Suite theme registry alter 'ds_entity_variables' not at the right place.

### ISSUE THREAD

**Issue by drupol** [drupol/registryonsteroids#124]
When using Display Suite, DS alter the registry and adds a callback: https://cgit.drupalcode.org/ds/tree/includes/ds.registry.inc#n266

The callback `ds_entity_variables` is not at the right place at the end of ROS mechanism.

The expected result of the **node** theme registry info (`$theme_registry['node']`) is :

```
Array
(
    [template] => node
    [path] => profiles/multisite_drupal_standard/themes/atomium/atomium/templates/node
    [type] => base_theme_engine
    [theme path] => sites/all/themes/ema
    [render element] => elements
    [preprocess functions] => Array
        (
            [0] => template_preprocess
            [1] => template_preprocess_node
            [2] => contextual_preprocess
            [3] => ctools_preprocess_node
            [4] => print_preprocess_node
            [5] => rdf_preprocess_node
            [6] => scheduler_preprocess_node
            [7] => views_preprocess_node
            [8] => entity_translation_preprocess_node
            [9] => field_group_build_entity_groups
            [10] => ds_entity_variables
            [11] => atomium_preprocess
            [12] => atomium_preprocess_node
            [13] => ec_europa_preprocess_node
            [14] => ema_preprocess_node
        )

    [process functions] => Array
        (
            [0] => template_process
            [1] => ctools_process
            [2] => rdf_process
            [3] => title_process_node
            [4] => atomium_process
        )
)
```

But the current result is:

```
Array
(
    [template] => node
    [path] => profiles/multisite_drupal_standard/themes/atomium/atomium/templates/node
    [type] => base_theme_engine
    [theme path] => sites/all/themes/ema
    [render element] => elements
    [preprocess functions] => Array
        (
            [0] => template_preprocess
            [1] => template_preprocess_node
            [2] => contextual_preprocess
            [3] => ctools_preprocess_node
            [4] => print_preprocess_node
            [5] => rdf_preprocess_node
            [6] => scheduler_preprocess_node
            [7] => views_preprocess_node
            [8] => entity_translation_preprocess_node
            [9] => atomium_preprocess
            [10] => atomium_preprocess_node
            [11] => ec_europa_preprocess_node
            [12] => ema_preprocess_node
            [17] => field_group_build_entity_groups
            [18] => ds_entity_variables
        )

    [process functions] => Array
        (
            [0] => template_process
            [1] => ctools_process
            [2] => rdf_process
            [3] => title_process_node
            [4] => atomium_process
        )

)
```

Same goes with the `field_group_build_entity_groups` callback, but I guess it's the same issue.

**Comment by donquixote** [comment]
Hi,
as I already said in the chat:
This is "works as designed".

### registry on steroids: "works as designed".

The module mimicks a situation where modules like DIsplay suite would add their own preprocess functions to each theme variant instead of just the base hook, in their hook_theme_registry_alter() implementation. In this situation, `ds_entity_variables()` would be added _after_ all discovered variant preprocessors.

The current behavior in registryonsteroids aims to be as close to Drupal core behavior as possible.

I think it is also quite similar to the preprocess order you would find in Drupal 8.

### Implications for Atomium

I think this is really a problem with Atomium, which previously caused a different preprocess order, where custom preprocess callbacks registered in hook_theme_registry_alter() were not in the place where one would expect them to be with core alone.

So, custom themes which relied on the old Atomium behavior now would need to be updated to the registryonsteroids behavior.

To avoid breaking these themes, Atomium could have its own atomium_theme_registry_alter, where it would modify the order of preprocess placeholders.

Before doing this, I strongly recommend to clearly define the intended order in Atomium, and compare it to core, and write unit tests.

**Comment by drupol** [comment]
First I'd like to say that I raise a false alarm with this issue.
There's nothing wrong with ROS and the way preprocess are ordered is perfectly fine.

I've spent my day trying to understand why the page was badly rendered on a client site and I made some discoveries.

I think it worth mentioning them here just in case if some people encounter some troubles later on.

Context:

- A Drupal 7 site
- Some content types
- Registry on steroids and Registry on steroids Alter enabled
- Atomium based theme with theme debug enabled
- Display Suite enabled.

Let's say we create a new node of type Article.
Let Display Suite manage the display, let's say we use the `DS 1 Col` layout preset.
Let's visit the page of the newly created node and display the source just before rendering the node:

```html
<!-- THEME DEBUG -->
  | <!-- CALL: theme('node__node__article__full') -->
  | <!-- FILE NAME SUGGESTIONS:
  | * node--node--article--full.tpl.php
  | * node--node--article.tpl.php
  | * node--node.tpl.php
  | * ds-1col--node--1.tpl.php
  | * ds-1col--node-article-full.tpl.php
  | * ds-1col--node-article.tpl.php
  | * ds-1col--node-full.tpl.php
  | * ds-1col--node.tpl.php
  | x ds-1col.tpl.php
  | * node--1.tpl.php
  | * node--article.tpl.php
  | * node.tpl.php
  | -->
  | <!-- BEGIN OUTPUT from 'sites/all/modules/contrib/ds/layouts/ds_1col/ds-1col.tpl.php' -->
  | <div  about="/node/1" typeof="sioc:Item foaf:Document" class="ds-1col node--node--article--full node-article node-promoted view-mode-full clearfix">...</div>
```

Nothing wrong there, we can see that the template used for displaying the node is `sites/all/modules/contrib/ds/layouts/ds_1col/ds-1col.tpl.php`.

Now, what if we want to override it ?
Let's create in the current theme a custom template that is available in the template suggestions: `node--node--article--full.tpl.php`

You will notice that the template won't be taken in account and you won't be able to override it, even if the call to `theme()` is using a variant.

I was almost sure that when you call `theme($hook)`, then the template `$hook.tpl.php` must be picked if exists, no matter what... but it's not the case when using Display Suite.
However, if a preprocess is created using that theme hook variant: `[HOOK]_preprocess_node__node__article__full(&$variables, $hook)`, it will be processed without any trouble.

If you disable the DS layout on the article content type, then, the custom template is chosen.

As this issue a big source of confusion for the clients (_the biggest?_), I think I will start to analyze how DS is working and see if we can find a way to have those template suggestions working as it should and maybe provide some fix/documentation.

I guess we can now close the issue as the original issue is actually not an issue but a configuration issue with Display Suite.
